app/javascript/mastodon/actions/compose.js

```diff
@@ -104,6 +108,24 @@ export function submitCompose() {
           dispatch(updateTimeline('public', { ...response.data }));
         }
       }
+
+      const statusTags = response.data.tags.map(it => it.name);
+      let tags = JSON.parse(localStorage.getItem('hash_tag_history'));
+      if (tags === null) {
+        tags = statusTags;
+      } else {
+        tags = tags.filter(it => !statusTags.includes(it));
+        tags.unshift(...statusTags);
+      }
+      const maxSize = 1000;
+      tags = tags.slice(0, maxSize);
+
+      const data = JSON.stringify(tags);
+      try {
+        localStorage.setItem('hash_tag_history', data);
+      } catch (e) {
+        //ignore
+      }
```

Comment:
Storing large amounts of data in LocalStorage is [inefficient](https://hacks.mozilla.org/2012/03/there-is-no-simple-solution-for-local-storage/). I'd advise using IDB instead via a library like [Dexie](http://dexie.org/). You can also use it for range queries, meaning you don't need to constantly re-parse a big string of JSON and then search through it using JavaScript.